src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -841,14 +787,106 @@ bool SotaUptaneClient::getNewTargets(std::vector<Uptane::Target> *new_targets, u
   return true;
 }
 
+std::unique_ptr<Uptane::Target> SotaUptaneClient::findTargetHelper(const Uptane::Targets &cur_targets,
+                                                                   const Uptane::Target &queried_target, int level,
+                                                                   bool terminating) {
+  auto it = std::find(cur_targets.targets.cbegin(), cur_targets.targets.cend(), queried_target);
+  if (it != cur_targets.targets.cend()) {
+    return std_::make_unique<Uptane::Target>(*it);
+  }
+
+  if (terminating || level >= Uptane::kDelegationsMaxDepth) {
+    return std::unique_ptr<Uptane::Target>(nullptr);
+  }
+
+  for (const auto &delegate_name : cur_targets.delegated_role_names_) {
+    Uptane::Role delegate_role = Uptane::Role::Delegation(delegate_name);
+    auto patterns = cur_targets.paths_for_role_.find(delegate_role);
+
+    if (patterns == cur_targets.paths_for_role_.end()) {
+      continue;
+    }
+
+    bool match = false;
+
+    for (const auto &pattern : patterns->second) {
+      if (fnmatch(pattern.c_str(), queried_target.filename().c_str(), 0) == 0) {
+        match = true;
+        break;
+      }
+    }
+
+    if (!match) {
+      continue;
+    }
+
+    // Target name matches one of the patterns
+
+    std::string delegation_meta;
+    std::shared_ptr<Uptane::Targets> delegation;
+
+    if (storage->loadDelegation(&delegation_meta, delegate_role)) {
+      delegation = Uptane::ImagesRepository::verifyDelegation(delegation_meta, delegate_role, cur_targets);
+      if (delegation == nullptr) {
+        storage->deleteDelegation(delegate_role);
+        delegation_meta.clear();
+      }
+    }
+
+    if (delegation == nullptr) {
+      if (!uptane_fetcher->fetchLatestRole(&delegation_meta, Uptane::kMaxImagesTargetsSize,
+                                           Uptane::RepositoryType::Image(), delegate_role)) {
+        throw std::runtime_error(std::string("Couldn't fetch ") +
+                                 delegate_role.ToString());  // TODO: connectivity exception
```

Comment:
Don't we still need to fetch even if we have a stored version in case the version has incremented? Maybe I'm confusing something with versions, but this seems like it won't ever fetch new versions.

Reply:
Oh wait, you're right. So storing delegations in a persistent storage doesn't make any sense at all, we don't know what's there until we go to the server. That's annoying.

Reply:
I can propose one of the two procedures then.

Reply:
Either:

1. At the beginning of update cycle clear all delegations.
2. Cache delegations during the update cycle.
3. Use cached version when it's available, download otherwise.

This way we can avoid downloading the same delegation over for different targets in the same update cycle, but still can't share them between the update cycles.

Reply:
Or:

1. Store delegations permanently.
2. Use the same procedure that is implemented now on the first pass.
3. If some a target has not been found, update delegations for it on a second pass.

This way we are more bandwith-effective, but there is a potential security problem in not being able to revoke trust to targets in delegated roles in a timely fashion.

Reply:
Second option doesn't seem safe enough to me, but the first option is interesting. Does it make since to do any sort of comparison between old and new versions? If we have both available anyway, I wonder if there's anything intelligent we can do with them or if we really just have to ignore the old one and throw it away.

Reply:
That's what's in docker notary documentation by the way:

> The snapshot key signs the snapshot metadata file, which enumerates the filenames, sizes, and hashes of the root, targets, and delegation metadata files for the collection.

Makes a lot of sense to me and resolves some of the points of confusion.

Reply:
This way we can cache the delegated meta and only update it if the version in snapshot.json has been incremented. And versioning for delegated targets suddenly makes sense again.

Reply:
That's fine, but is that what the backend is doing? I also want to make sure TUF/Uptane are friendly to that notion, but I think they should be. To make sure I understand correctly: this seems like the second option you mentioned, but more robust, because we don't store delegations "permanently", just as long as the version specified in snapshot.json hasn't changed.

Reply:
Yes. TUF spec says

> The snapshot.json file is signed by the snapshot role. It lists the version numbers of all metadata on the repository, excluding timestamp.json and mirrors.json. For the root role, the hash(es), size, and version number are listed.

which doesn't contradict to what's in Notary docs, just does not name delegations explicitly. I believe that if backend doesn't add delegated roles to snapshot.json yet, it should do so, because otherwise we have weaker security guarantees for delegated targets than for top-level target, and I can't see a reason why it should be the case.